tests/utils/exec.ts

```diff
@@ -180,219 +198,134 @@ async function pkStdio(
-        if (error != null && error.code === undefined) {
-          // This can only happen when the command is killed
-          return reject(error);
-        } else {
-          // Success and Unsuccessful exits are valid here
-          return resolve({
-            exitCode: error && error.code != null ? error.code : 0,
-            stdout,
-            stderr,
-          });
-        }
-      },
-    );
-  });
+  if (opts.command == null) {
+    return await pkExecWithoutShell(args, opts);
+  } else {
+    return await pkExecWithShell(args, opts);
+  }
 }
 
 /**
  * Launch pk command through subprocess
  * This is used when a subprocess functionality needs to be used
  * This is intended for non-terminating subprocesses
- * @param env Augments env for command execution
- * @param cwd Defaults to temporary directory
+ * By default `globalThis.testCommand` should be `undefined` because `PK_TEST_COMMAND` will not be set
+ * This is strictly checking for existence, `PK_TEST_COMMAND=''` is legitimate but undefined behaviour
  */
 async function pkSpawn(
   args: Array<string> = [],
-  env: Record<string, string | undefined> = {},
-  cwd?: string,
+  opts: ExecOpts = { env: {}, command: globalThis.testCmd },
   logger: Logger = new Logger(pkSpawn.name),
 ): Promise<ChildProcess> {
-  if (globalThis.testCmd != null) return pkSpawnTarget(args, env, cwd, logger);
-
-  cwd =
-    cwd ??
-    (await fs.promises.mkdtemp(path.join(globalThis.tmpDir, 'polykey-test-')));
-  env = {
-    ...process.env,
-    ...env,
-  };
-  // Recall that we attempt to connect to all specified seed nodes on agent start.
-  // Therefore, for testing purposes only, we default the seed nodes as empty
-  // (if not defined in the env) to ensure no attempted connections. A regular
-  // PolykeyAgent is expected to initially connect to the mainnet seed nodes
-  env['PK_SEED_NODES'] = env['PK_SEED_NODES'] ?? '';
-  const tsConfigPath = path.resolve(
-    path.join(globalThis.projectDir, 'tsconfig.json'),
-  );
-  const polykeyPath = path.resolve(
-    path.join(globalThis.projectDir, 'src/bin/polykey.ts'),
-  );
-  const command =
-    globalThis.testCmd != null
-      ? path.resolve(path.join(globalThis.projectDir, globalThis.testCmd))
-      : 'ts-node';
-  const tsNodeArgs =
-    globalThis.testCmd != null ? [] : ['--project', tsConfigPath, polykeyPath];
-  const subprocess = child_process.spawn(command, [...tsNodeArgs, ...args], {
-    env,
-    cwd,
-    stdio: ['pipe', 'pipe', 'pipe'],
-    windowsHide: true,
-  });
-  // The readline library will trim newlines
-  const rlOut = readline.createInterface(subprocess.stdout!);
-  rlOut.on('line', (l) => logger.info(l));
-  const rlErr = readline.createInterface(subprocess.stderr!);
-  rlErr.on('line', (l) => logger.info(l));
-  return subprocess;
+  if (opts.command == null) {
+    return await pkSpawnWithoutShell(args, opts, logger);
+  } else {
+    return await pkSpawnWithShell(args, opts, logger);
+  }
```

Comment:
This seems odd to me. We have two commands here that do mostly the same thing. Wouldn't it be cleaner if we had 1 more generic spawn command that can take the command, args, opts and logger as parameters. Then here we can pass in the needed command and augment the augments depending if opts.command is set?

The less we duplicate code the better but I'm not sure if it's worth changing at this point. Thoughts?

Reply:
If this is the worker and wrapper pattern then it's fine to do this.

Reply:
If there are no catch handler, it's fine to `return pkSpawn...`, no need to do an intermediate `await`.

Reply:
Removed the intermediate awaits from `pkExec` and `pkSpawn`

Reply:
Resolving because this is a fine way of doing worker and wrapper pattern. Just remember to export the `pkSpawnWithoutShell` and `pkSpawnWithShell` as well.